### PR TITLE
[10.x] Precognition

### PR DESCRIPTION
**Pull request by timacdonald**
See: https://github.com/laravel/framework/pull/44339

### CODE REVIEW

**Comment by phh** [comment]
@timacdonald This looks awesome! We hope to adopt this as soon as possible! Any plans to when to release the npm packages - or if you would make the repositories public while they're getting worked on? 🙏

**Comment by mreduar** [comment]
> @timacdonald This looks awesome! We hope to adopt this as soon as possible! Any plans to when to release the npm packages - or if you would make the repositories public while they're getting worked on? 🙏

I would also love to try it, I have been following the PR's since Laracon 😅

File: precognition.md [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+    return to_route('users.show', ['user' => $user]);
+})->middleware(HandlePrecognitiveRequests::class);
+```
+
+When a precognitive request hits this route all middleware will run, the form request (or any other route dependencies) will be resolved and the form request's validation will run. After validation, regardless of if the validation passed or failed, the controller will not be executed.
+
+<a name="handling-precognitive-requests"></a>
+### Handling Precognitive Requests
+
+Precognitive requests should generally be side-effect free. This is where the Precognition "pattern" comes in. It is recommend when you add Precognition to your routes that you consider the side-effects triggered in your application's middleware and form requests. If the side-effects present should be skipped for precognitive requests, you may need to add a conditional around the side-effect.
```
Comment (mreduar) [inline precognition.md:69]:
```suggestion
Precognitive requests should generally be side-effect free. This is where the Precognition "pattern" comes in. It is recommended when you add Precognition to your routes that you consider the side-effects triggered in your application's middleware and form requests. If the side-effects present should be skipped for precognitive requests, you may need to add a conditional around the side-effect. 
```

File: precognition.md [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+        <InputLabel for="username" />
+        <TextInput id="username" v-model="form.username" />
+        <InputError :message="errors.username?.[0]" />
+
+        <!-- ... -->
+    </form>
+</template>
+```
+
+We will enhance this implementation by adding live validation powered by Laravel Precognition. To achieve this we will create a precognitive form, passing through the method, url, and initial form data. Then we will:
```
Comment (mreduar) [inline precognition.md:598]:
```suggestion
We will enhance this implementation by adding live validation powered by Laravel Precognition. To achieve this, we will create a precognitive form, passing through the method, url, and initial form data. Then we will:
```

File: precognition.md [hunk 3]
```diff
@@ -0,0 +1,837 @@
...
+- `validating`: The latest input name awaiting.
+
+Also see: [configuring the Validator](#vue-validation-configuration).
+
+<a name="specification"></a>
+## Specification
+
+- Precognition requests MUST have a `Precognition` header with the value `true`.
+- Precognition responses MUST have a `Precognition` header with the value `true`.
+- When requesting specific inputs be validated, the `Precognition-Validate-Only` header SHOULD be used. The value MUST be a comma seperated list in input names e.g. `email,phone,address`.
```
Comment (mreduar) [inline precognition.md:835]:
```suggestion
- When requesting specific inputs be validated, the `Precognition-Validate-Only` header SHOULD be used. The value MUST be a comma separated list in input names e.g. `email,phone,address`.
```

File: precognition.md [hunk 4]
```diff
@@ -0,0 +1,837 @@
...
+precognitive.post('/projects/5', data, {
+    fingerprint: 'request-1',
+});
+
+precognitive.post('/projects/5', data, {
+    fingerprint: 'request-2',
+});
+```
+
+To disable this feature you may specify a fingerprint of `null`. You may disable it globally by passing `null` to `fingerprintRequestsUsing`:
```
Comment (mreduar) [inline precognition.md:315]:
```suggestion
To disable this feature, you may specify a fingerprint of `null`. You may disable it globally by passing `null` to `fingerprintRequestsUsing`:
```

File: precognition.md [hunk 5]
```diff
@@ -0,0 +1,837 @@
...
+- Live validation of forms, powered by Laravel validation rules.
+- Notifying users that a resource they are editing has been updated since it was retrieved.
+- Notifying users their session has expired.
+
+Precognition works by executing all middleware and resolving all controller dependencies (including form requests) of a particular route, but not executing the route's controller. You will also see that Precognition is part feature and part pattern.
+
+<a name="installation"></a>
+## Installation
+
+The frontend helper libraries make working with Precognition a dreamy delight. If you are going to use Precognition, we recommend installing the appropriate library for your project. There is a vanilla JavaScript and Vue flavoured package available via NPM:
```
Comment (mreduar) [inline precognition.md:32]:
```suggestion
The frontend helper libraries make working with Precognition a dreamy delight. If you are going to use Precognition, we recommend installing the appropriate library for your project. There is a vanilla JavaScript and Vue flavored package available via NPM:
```

File: precognition.md [hunk 6]
```diff
@@ -0,0 +1,837 @@
...
+    headers: { 'X-Requested-With': 'XMLHttpRequest' }
+});
+
+precognitive.use(client);
+```
+
+<a name="configuration"></a>
+### Configuration
+
+The Precognition configuration object is the [Axios' configuration](https://axios-http.com/docs/req_config) object with some additional options to customize behaviour and also make handling common responses easier.
```
Comment (mreduar) [inline precognition.md:350]:
```suggestion
The Precognition configuration object is the [Axios' configuration](https://axios-http.com/docs/req_config) object with some additional options to customize behavior and also make handling common responses easier.
```

File: precognition.md [hunk 7]
```diff
@@ -0,0 +1,837 @@
...
+If you would like to change the default debounce timeout for validation requests, you may use the form's `setValidationTimeout` function:
+
+```js
+form.setValidationTimeout({ seconds: 2 });
+```
+
+<a name="validating-vue-inertia"></a>
+### Working With Vue and Inertia
+
+Inertia has a built-in form helper that makes working with forms a lovely experience. We wanted to maintain this experience while enabling realtime validation with Precognition, so we decided to wrap Inertia's form helper.
```
Comment (vurpa) [inline precognition.md:716]: Any plans to support React too?

**Comment by StubyProjects** [comment]
Hey, pretty cool feature so far. One question. How do I validate nested fields in an array. E.g calling form.validate('array.0.email') with inertia is not validated currently.

**Comment by panda-madness** [comment]
Awesome work guys! Any idea when the docs are going to get merged? Or do you want to finish up the JS packages before that?

**Comment by ahamilton9** [comment]
Any idea what happened? It looks like the backend code got merged into v9, but the docs and JS packages were incomplete. v10 just launched the other day. The Vue npm package points to github.com/laravel/precognition which doesn't seem to exist.

**Comment by ProjectAsphaleia** [comment]
> Any idea what happened? It looks like the backend code got merged into v9, but the docs and JS packages were incomplete. v10 just launched the other day. The Vue npm package points to github.com/laravel/precognition which doesn't seem to exist.

I was using Precognition in one of my projects.
Sadly got the following error when I was building my app with SSR. 
> Error [ERR_MODULE_NOT_FOUND]: Cannot find module '..\node_modules\laravel-precognition-vue\dist\inertia' imported from ..\node_modules\laravel-precognition-vue\dist\index.js

Found out after that the`laravel-precognition` and `laravel-precognition-vue` are gone. Not to sure what happened with it, probably got set to private. Hope it gets added soon again, it's an amazing feature.

@timacdonald Any insights on where it went and when we can expect it to "release" again?

**Comment by nwehrhan** [comment]
I was following precognition for awhile? Anyone know the state it's in?

**Comment by mreduar** [comment]
> I was following precognition for awhile? Anyone know the state it's in?

I have installed it and tested it by doing a hack or two, but it works. But I still notice that it lacks features.

**Comment by nwehrhan** [comment]
@mreduar  What features is it lacking?

**Comment by mreduar** [comment]
> @mreduar What features is it lacking?

One of them is to be able to validate more than one input at the same time, for now I could not get documentation nor seeing the code how to validate more than one input.

**Comment by driesvints** [comment]
Going to close this since it's targetting the old 9.x branch. Feel free to reattempt this against 10.x, thanks.

**Comment by driesvints** [comment]
@timacdonald should this PR be reconsidered?

**Comment by timacdonald** [comment]
@driesvints we are discussing it internally at the moment.

**Comment by lazer-hybiz** [comment]
any updates?

File: 10.x [hunk 8]
```diff
@@ -0,0 +1 @@
+10.x
```
Comment (driesvints) [inline 10.x:1]: @timacdonald this seems accidentally?

**Comment by timacdonald** [comment]
_"For the record, we never broke up; we just took a fourteen-year vacation"._

**Comment by lazer-hybiz** [comment]
@timacdonald any chance the frontend packages repo will be made public?

**Pull request merged**